ngraph/core/src/op/interpolate.cpp

```diff
@@ -111,143 +114,538 @@ namespace ngraph
+        auto pads_end = correct_pad(attrs.pads_end, input_rank);
+
+        auto padded_input_shape_vector = get_padded_input_shape(input_shape, pads_begin, pads_end);
+        std::vector<std::size_t> out_shape_vector;
+
+        if (attrs.shape_calculation_mode == ShapeCalcMode::scales)
+        {
+            out_shape_vector = shape_infer_with_scales(padded_input_shape_vector, axes, scales);
+        }
+        else
+        {
+            auto target_shape = get_target_shape_vector(args, num_of_axes);
+            out_shape_vector =
+                shape_infer_with_target_shape(padded_input_shape_vector, axes, target_shape);
+        }
+
+        Shape out_shape{out_shape_vector};
+        Shape padded_input_shape{padded_input_shape_vector};
+
+        return {input_shape, padded_input_shape, out_shape, pads_begin, pads_end, axes, scales};
+    }
+
+    template <element::Type_t ET>
+    inline bool evaluate(const HostTensorVector& args,
+                         const HostTensorPtr& out,
+                         const op::v4::Interpolate::InterpolateAttrs& attrs)
+    {
+        using T = typename element_type_traits<ET>::value_type;
+
+        auto info_for_reference = get_info_to_call_reference(args, attrs);
+
+        out->set_element_type(args[0]->get_element_type());
+        out->set_shape(info_for_reference.out_shape);
+
+        std::vector<T> padded_input_data(shape_size(info_for_reference.padded_input_shape), T{});
+
+        CoordinateTransform input_transform(info_for_reference.input_shape);
+        CoordinateTransform padded_transform(info_for_reference.padded_input_shape);
+
+        const T* data_ptr = args[0]->get_data_ptr<ET>();
+        T* padded_data_ptr = padded_input_data.data();
+
+        for (const Coordinate& input_coord : input_transform)
+        {
+            auto padded_coord = input_coord;
+            std::size_t i = 0;
+            for (std::size_t pad : info_for_reference.pads_begin)
+            {
+                padded_coord[i] += pad;
+            }
+            padded_data_ptr[padded_transform.index(padded_coord)] =
+                data_ptr[input_transform.index(input_coord)];
+        }
+
+        runtime::reference::interpolate<T>(padded_input_data.data(),
+                                           info_for_reference.padded_input_shape,
+                                           info_for_reference.scales,
+                                           info_for_reference.axes,
+                                           out->get_data_ptr<ET>(),
+                                           info_for_reference.out_shape,
+                                           attrs);
+        return true;
+    }
+
+    bool evaluate_interpolate_v4(const HostTensorVector& args,
+                                 const HostTensorPtr& out,
+                                 const op::v4::Interpolate::InterpolateAttrs& attrs)
+    {
+        bool rc = true;
+        switch (args[0]->get_element_type())
```

Comment:
You can remove this switch case if evaluate method doesn't have template arguments.

Reply:
We cannot remove this switch case, because Interpolate-4 specification specifies `fp16`, `fp32`, `int8` as admissible types of elements of an input tensor.

Reply:
`validate_and_infer_types` (not evaluate method) should check available arguments.

`evaluate` method can work only on data types which were validated by the `validate_and_infer` method.

**It means that switch case and template arguments are redundant!**

I don't want to reduce the body of a template `evaluate` method, I want to remove template arguments for `evaluate` method. Template arguments are not required for this method.
You can add assert if you want to recheck supported data_types, but in any case we don't need template arguments and template code. `HostTensor` contains all information which is needed for it.